### ISSUE TITLE
Advertising specific GPU types as separate extended resource

### ISSUE THREAD

**Issue by deepanker-s** [NVIDIA/k8s-device-plugin#424]
Hello,
I am working at Uber.

### 1. Feature description
Advertising special hardware (specific GPU types say A100) as a separate extended resource.

As of now, we have a blanket of "nvidia.com/gpu" for all types of GPUs that this plugin supports. If we want our pods to run specifically on some GPU types, then, we need to be able to request such a resource.

For requesting such a specific resource, there are 2 ways -
1. [Existing] Using nodeLabels/nodeSelectors
2. [New] Advertising the same directly as a new resource such as "nvidia.com/gpu-A100-...."

This added functionality can be enabled based upon a configuration flag and can use [gpu-feature-discovery](https://github.com/NVIDIA/gpu-feature-discovery) labels to extract the SKU/GPU type.

### 2. Why
1. There is already a similar resource advertising being done for MIG enabled devices -
```
nvidia.com/gpu
nvidia.com/mig-1g.5gb
nvidia.com/mig-2g.10gb
nvidia.com/mig-3g.20gb
nvidia.com/mig-7g.40gb
```

2. Another reason is that usage of nodeLabels/nodeSelectors may not be possible due to some limitations.

### 3. Similar existing work
I found a [design doc](https://docs.google.com/document/d/1dL67t9IqKC2-xqonMi6DV7W2YNZdkmfX7ibB6Jb-qmk/edit#heading=h.jw5js7865egx) for "Custom Resource Naming and Supporting Multiple GPU SKUs on a Single Node in Kubernetes".

It is actually advertising different types of GPUs as new resource name but those different GPU cards should be on the same node. I am not sure whether the same will also support if the corresponding GPU cards/types are on different nodes as well.

### 4. Summary of queries
1. Is the above feature request already being supported by the above mentioned "Similar existing work"?
2. If yes, when will that work be approved and available?

**Comment by klueska** [comment]
There is still no planned support for this in the k8s-device plugin. All of the functionality is there (as described in the link you provided), but it is explicitly disabled by this line in the code https://github.com/NVIDIA/k8s-device-plugin/blob/main/cmd/nvidia-device-plugin/main.go#L239.

The future for supporting multiple GPU cards per node is via a new mechanism in Kubernetes called Dynamic Resource Allocation (DRA):
https://docs.google.com/document/d/1BNWqgx_SmZDi-va_V31v3DnuVwYnF2EmN7D-O_fB6Oo/edit
https://github.com/NVIDIA/k8s-dra-driver

**Comment by deepanker-s** [comment]
Hey Kevin,
Thanks for the info.

I was actually asking about specific GPU resource naming for GPUs on different nodes (not on same node).
But looks like, the answer seems to be the same. DRA can help achieve that as well.

**Comment by deepanker-s** [comment]
Hey Kevin,
I understand now that DRA can be used to specify GPU types (A100, H100) for different pods using "GpuClaimParameters".

Is there any functionality to advertise these specified resources?

Example -
Using DRA "GpuClaimParameters") (as in [gpu-test6 example](https://gitlab.com/nvidia/cloud-native/k8s-dra-driver/-/blob/main/demo/gpu-test6.yaml)), if -
* podA is scheduled on A100 GPU
* podB is scheduled on H100 GPU

Will device plugin advertise the resource usage details - how many A100 devices are being used? 
Like currently we adverstise - 
`nvidia.com/gpu : 10`

Will it provide details such as below in any manner?
`nvidia.com/gpu-A100 : 5`

**Comment by dimm0** [comment]
We're looking to install the [yunikorn](https://yunikorn.apache.org/) scheduler on the cluster and having different resources for different GPUs will help a lot to prioritize the use of more powerful (and less available) GPUs among users using the fair share. It's impossible to do with just labels.

**Comment by sjdrc** [comment]
> There is still no planned support for this in the k8s-device plugin
Is there a reason why this isn't planned to be implemented here? This seems like an essential feature for any cluster with more than 1 model of GPU and there's currently no adequate workaround at all.

**Comment by klueska** [comment]
It was a product decision, not an engineering one.

All of the code to support it is merged in the plugin and simply disabled by https://github.com/NVIDIA/k8s-device-plugin/blob/main/cmd/nvidia-device-plugin/main.go#L239.

The decision not to support this gets revisited periodically, but our product team is still not in favor of it, so our hands are tied.

If you want to enable it in a custom build of the plugin, just remove that line referenced above and it should work as described in https://docs.google.com/document/d/1dL67t9IqKC2-xqonMi6DV7W2YNZdkmfX7ibB6Jb-qmk/edit#heading=h.jw5js7865egx.

**Comment by yuzliu** [comment]
@klueska thanks for the explanation. We also explored the extended resource options and we even have a component we wrote ourselves to patch node with gpu extended resources. Just curious would you be open to add a flag to turn this feature on/off so we don't have to deploy a customized version of nvidia device plugin?

**Comment by klueska** [comment]
@yuzliu Do you have multiple GPU types per node? If not, are [node-labels from GFD](https://github.com/NVIDIA/gpu-feature-discovery) / nodeSelectors not enough for your use case?

**Comment by yuzliu** [comment]
@klueska Thanks for the reply! We don't have multiple GPU types per node but we do have multiple GPU types per cluster. We have already deployed the GPU feature discovery and have gpu product label on each GPU node but that doesn't solve our problem because:
1. We have clusters having multiple GPU types e.g. A100 + T4 mixed in one cluster
2. We have resourcequota on each namespace and we want to achieve resource quota enforcement e.g. namespace A can only use 1 A100 and 5 T4 on a namespace level.
3. We want to collect metrics accurately per each GPU type. For example we'd like to know namespace A has 4 A100 available and 1 A100 was requested but still have 3 left.

**Comment by klueska** [comment]
Got it -- labels from GPU feature discovery are sufficient for 1, but not for 2 and 3 -- for that you need a unique extended resources.

**Comment by yuzliu** [comment]
Yep, we even have an internal component to advertise extended resources e.g. V100, A100 and T4. But I'd really love to have less customized logic internally but rely on Nvidia's official component to make our long term maintenance easier.

**Comment by klueska** [comment]
As mentioned before -- it is not up to me. Product Management is still against it, so for the time being the answer is no.

**Comment by leoncamel** [comment]
Any progress on this issue?

**Comment by ZDWWWWW** [comment]
Any progress on this issue?

**Comment by csh0101** [comment]
> We're looking to install the [yunikorn](https://yunikorn.apache.org/) scheduler on the cluster and having different resources for different GPUs will help a lot to prioritize the use of more powerful (and less available) GPUs among users using the fair share. It's impossible to do with just labels.我们正在考虑在集群上安装 [yunikorn](https://yunikorn.apache.org/) 调度程序，为不同的 GPU 分配不同的资源将有助于在使用公平共享的用户中优先使用性能更强大（且可用性更低）的 GPU。仅靠标签是无法做到这一点的。

why don't use volcano ?

**Comment by dimm0** [comment]
How would volcano track different GPUs? Can it do by node label?

**Comment by csh0101** [comment]
> How would volcano track different GPUs? Can it do by node label?Volcano 如何追踪不同的 GPU？

yes volcano fork the nvidia/device and  enable the classifer gpu type.  I do it in my lab env.. 

and node label is offer by nfd . standrad method to support

**Comment by csh0101** [comment]
> Got it -- labels from GPU feature discovery are sufficient for 1, but not for 2 and 3 -- for that you need a unique extended resources.明白了——GPU 功能发现的标签对于 1 来说是足够的，但对于 2 和 3 来说则不够，你需要一个独特的扩展资源。

yes， some oss project like volcano use a queue to limit the usage of resource , it's need different resource name . therefore, one uniform name is not appericate

**Comment by wkd-woo** [comment]
+1

any update about this issue?

**Comment by shivamerla** [comment]
As @klueska [commented](https://github.com/NVIDIA/k8s-device-plugin/issues/424#issuecomment-1647438332) this can be natively achieved today using DRA (which is GA from K8s 1.34) using CEL expressions in the device requests within a resource claim. Please refer to the examples [here](https://github.com/kubernetes-sigs/dra-driver-nvidia-gpu/tree/main/demo).
